File: src/strategy_v1/HemaTrendLevelConfirmationStrategy.js
import { HemaTrendLevelStrategyV2 } from './HemaTrendLevelStrategyV2';

/**
 * HEMA Trend Level Confirmation Strategy
 * Extends HemaTrendLevelStrategyV2 with additional confirmation requirements
 */
export class HemaTrendLevelConfirmationStrategy extends HemaTrendLevelStrategyV2 {
    constructor(config = {}) {
        super(config);
        this.confirmationBars = config.confirmationBars || 2;
    }

    /**
     * Confirm bullish trend with price action and momentum
     */
    confirmBullishTrend(candleData, hemaFast) {
        const lastIdx = candleData.length - 1;
        const lastClose = candleData[lastIdx].close;
        const prevClose = candleData[lastIdx - 1].close;
        
        const lastHemaFast = hemaFast[hemaFast.length - 1];
        const prevHemaFast = hemaFast[hemaFast.length - 2];

        // Price confirmation: Close above fast HEMA
        const priceConfirm = lastClose > lastHemaFast;
        
        // Momentum confirmation: Fast HEMA rising
        const momentumConfirm = lastHemaFast > prevHemaFast;
        
        return priceConfirm && momentumConfirm;
    }

    /**
     * Confirm bearish trend with price action and momentum
     */
    confirmBearishTrend(candleData, hemaFast) {
        const lastIdx = candleData.length - 1;
        const lastClose = candleData[lastIdx].close;
        const prevClose = candleData[lastIdx - 1].close;
        
        const lastHemaFast = hemaFast[hemaFast.length - 1];
        const prevHemaFast = hemaFast[hemaFast.length - 2];

        // Price confirmation: Close below fast HEMA
        const priceConfirm = lastClose < lastHemaFast;
        
        // Momentum confirmation: Fast HEMA falling
        const momentumConfirm = lastHemaFast < prevHemaFast;
        
        return priceConfirm && momentumConfirm;
    }

    /**
     * Analyze with confirmation
     */
    analyze(candleData) {
        if (candleData.length < Math.max(this.fastLen, this.slowLen) + this.confirmationBars + 20) {
            return { signal: 'NONE', message: '', confidence: 0 };
        }

        const hemaIndicators = this.applyHemaIndicators(candleData);
        const lastIdx = candleData.length - 1;
        const crossIdx = hemaIndicators.bullishCross.length - 1;
        const trendIdx = hemaIndicators.trend.length - 1;

        // Bullish cross with confirmation
        if (hemaIndicators.bullishCross[crossIdx]) {
            if (this.confirmBullishTrend(candleData, hemaIndicators.hemaFast)) {
                return this.createSignal(
                    candleData,
                    'BUY',
                    `HEMA bullish cross with confirmation: Fast(${this.fastLen}) > Slow(${this.slowLen})`,
                    0.85
                );
            }
        }

        // Bearish cross with confirmation
        if (hemaIndicators.bearishCross[crossIdx]) {
            if (this.confirmBearishTrend(candleData, hemaIndicators.hemaFast)) {
                return this.createSignal(
                    candleData,
                    'SELL',
                    `HEMA bearish cross with confirmation: Fast(${this.fastLen}) < Slow(${this.slowLen})`,
                    0.85
                );
            }
        }

        // Bull support test with confirmation
        if (hemaIndicators.trend[trendIdx] === 1 && hemaIndicators.bullTest[crossIdx]) {
            const prevClose = candleData[lastIdx - 1].close;
            const lastClose = candleData[lastIdx].close;
            
            if (this.confirmBullishTrend(candleData, hemaIndicators.hemaFast) && lastClose > prevClose) {
                return this.createSignal(
                    candleData,
                    'BUY',
                    'Bull support test with confirmation in uptrend',
                    0.8
                );
            }
        }

        // Bear resistance test with confirmation
        if (hemaIndicators.trend[trendIdx] === -1 && hemaIndicators.bearTest[crossIdx]) {
            const prevClose = candleData[lastIdx - 1].close;
            const lastClose = candleData[lastIdx].close;
            
            if (this.confirmBearishTrend(candleData, hemaIndicators.hemaFast) && lastClose < prevClose) {
                return this.createSignal(
                    candleData,
                    'SELL',
                    'Bear resistance test with confirmation in downtrend',
                    0.8
                );
            }
        }

        return { signal: 'NONE', message: '', confidence: 0 };
    }
}

File: python_strategy/HemaTrendLevelConfirmationStrategy.py
import os
import logging
from pandas import DataFrame

from entity import EventType
from common import CLOSE
from .HemaTrendLevelStrategyV2 import HemaTrendLevelStrategy
from .StopTargetEnum import StopTargetEnum

from common import get_logger
logger = get_logger(__name__)

class HemaTrendLevelConfirmationStrategy(HemaTrendLevelStrategy):
    def __init__(self, fast_len: int = 11, slow_len: int = 21, adx_period: int = 14,
                 adx_threshold: int = 20, atr_period: int = 14,
                 sl_target: StopTargetEnum = StopTargetEnum.SL_50_200,
                 confirmation_bars: int = 2):
        super().__init__(fast_len, slow_len, adx_period, adx_threshold, atr_period, sl_target)
        self.confirmation_bars = confirmation_bars

    def analyze(self, df: DataFrame) -> tuple[EventType, str]:
        if len(df) < max(self.fast_ema, self.slow_ema) + self.confirmation_bars:
            return EventType.NONE, ""

        self.apply_indicator(df)
        last = df.iloc[-1]
        prev = df.iloc[-2]

        # Bullish cross with confirmation
        if last['bullish_cross']:
            if self._confirm_bullish_trend(df):
                msg = f"HEMA bullish cross with confirmation: Fast({self.fast_ema}) > Slow({self.slow_ema})"
                logger.info(msg)
                return EventType.BUY, msg

        # Bearish cross with confirmation
        if last['bearish_cross']:
            if self._confirm_bearish_trend(df):
                msg = f"HEMA bearish cross with confirmation: Fast({self.fast_ema}) < Slow({self.slow_ema})"
                logger.info(msg)
                return EventType.SELL, msg

        # Bull support test with confirmation
        if last['trend'] == 1 and last['bull_test']:
            if self._confirm_bullish_trend(df) and prev['trend'] == 1 and last[CLOSE] > prev[CLOSE]:
                msg = f"Bull support test with confirmation in uptrend"
                logger.info(msg)
                return EventType.BUY, msg

        # Bear resistance test with confirmation
        if last['trend'] == -1 and last['bear_test']:
            if self._confirm_bearish_trend(df) and prev['trend'] == -1 and last[CLOSE] < prev[CLOSE]:
                msg = f"Bear resistance test with confirmation in downtrend"
                logger.info(msg)
                return EventType.SELL, msg

        return EventType.NONE, ""

    def _confirm_bullish_trend(self, df: DataFrame) -> bool:
        """Confirm bullish signal with price action and momentum"""
        last = df.iloc[-1]
        prev = df.iloc[-2]
        
        # Price confirmation: Close above fast HEMA
        price_confirm = last[CLOSE] > last[self.hema_fast_column]
        
        # Momentum confirmation: Fast HEMA rising
        momentum_confirm = last[self.hema_fast_column] > prev[self.hema_fast_column]
        
        return price_confirm and momentum_confirm

    def _confirm_bearish_trend(self, df: DataFrame) -> bool:
        """Confirm bearish signal with price action and momentum"""
        last = df.iloc[-1]
        prev = df.iloc[-2]
        
        # Price confirmation: Close below fast HEMA
        price_confirm = last[CLOSE] < last[self.hema_fast_column]
        
        # Momentum confirmation: Fast HEMA falling
        momentum_confirm = last[self.hema_fast_column] < prev[self.hema_fast_column]
        
        return price_confirm and momentum_confirm
